packages/@functionless/formation/src/aws.ts

```diff
@@ -0,0 +1,109 @@
+import { Account, ClientOptions, IAws } from "cdk-assets";
+import s3_v2 from "aws-sdk/clients/s3";
+import secrets_manager_v2 from "aws-sdk/clients/secretsmanager";
+import ecr_v2 from "aws-sdk/clients/ecr";
+import sts from "@aws-sdk/client-sts";
+import * as os from "os";
+import { ChainableTemporaryCredentials, Credentials } from "aws-sdk";
+
+export default class AwsClient implements IAws {
+  constructor(
+    private account: string,
+    private region: string,
+    private sdkConfig?: any
+  ) {}
+
+  async discoverPartition(): Promise<string> {
+    return "aws";
+  }
+  async discoverDefaultRegion(): Promise<string> {
+    return this.region;
+  }
+  async discoverCurrentAccount(): Promise<Account> {
+    return {
+      accountId: this.account,
+      partition: await this.discoverPartition(),
+    };
+  }
+  async discoverTargetAccount(options: ClientOptions): Promise<Account> {
+    const stsClient = await this.stsClient(await this.awsOptions(options));
+    const response = await stsClient.send(new sts.GetCallerIdentityCommand({}));
+    if (!response.Account || !response.Arn) {
+      throw new Error(
+        `Unrecognized response from STS: '${JSON.stringify(response)}'`
+      );
+    }
+    return {
+      accountId: response.Account!,
+      partition: response.Arn!.split(":")[1]!,
+    };
+  }
+  async s3Client(options: ClientOptions): Promise<s3_v2> {
+    return new s3_v2(options);
+  }
+  async stsClient(options: ClientOptions): Promise<sts.STSClient> {
+    return new sts.STSClient(options);
+  }
+  async ecrClient(options: ClientOptions): Promise<ecr_v2> {
+    return new ecr_v2(options);
+  }
+  async secretsManagerClient(
+    options: ClientOptions
+  ): Promise<secrets_manager_v2> {
```

Comment:
Should these be memoized ?

Reply:
I suspect CDK does memoize them internally, but new options can be passed in on each call, so I didn't memorize them.